google-cloud-logging/src/main/java/com/google/cloud/logging/LogEntry.java

```diff
@@ -575,6 +583,145 @@ public Builder toBuilder() {
+     * @param value an object to be serialized to Json using {@link Gson}.
+     * @param appendComma a flag to add a trailing comma.
+     * @return a reference to this object.
+     */
+    public StructuredLogFormatter appendField(String name, Object value, boolean appendComma) {
+      checkNotNull(name);
+      if (value != null) {
+        builder.append(gson.toJson(name)).append(":").append(gson.toJson(value));
+        if (!appendComma) {
+          builder.append(",");
+        }
+      }
+      return this;
+    }
+
+    public StructuredLogFormatter appendField(String name, Object value) {
+      return appendField(name, value, false);
+    }
+
+    /**
+     * Adds a collection of Json fields that {@code value} parameter is serialized to to the current
+     * string representation. Nothing is added if {@code value} parameter is {@code null}.
+     *
+     * @param value an object to be serialized to Json using {@link Gson}.
+     * @param appendComma a flag to add a trailing comma.
+     * @return a reference to this object.
+     */
+    public StructuredLogFormatter appendJson(Object value, boolean appendComma) {
+      if (value != null) {
+        String json = gson.toJson(value);
+        // append json object without brackets
+        if (json.length() > 1) {
+          builder.append(json.substring(0, json.length() - 1).substring(1));
+          if (!appendComma) {
+            builder.append(",");
+          }
+        }
+      }
+      return this;
+    }
+
+    public StructuredLogFormatter appendJson(Object value) {
+      return appendJson(value, false);
+    }
+  }
+
+  /**
+   * Serializes the object to a one line JSON string in the simplified format that can be parsed by
+   * the logging agents that run on Google Cloud resources.
+   */
+  public String toStructuredJsonString() {
+    final StringBuilder builder = new StringBuilder("{");
+    final StructuredLogFormatter formatter = new StructuredLogFormatter(builder);
+
+    formatter
+        .appendField("severity", severity)
+        .appendField("timestamp", timestamp)
+        .appendField("httpRequest", httpRequest)
+        .appendField("logging.googleapis.com/insertId", insertId)
+        .appendField("logging.googleapis.com/labels", labels)
+        .appendField("logging.googleapis.com/operation", operation)
+        .appendField("logging.googleapis.com/sourceLocation", sourceLocation)
+        .appendField("logging.googleapis.com/spanId", spanId)
+        .appendField("logging.googleapis.com/trace", trace)
+        .appendField("logging.googleapis.com/trace_sampled", traceSampled);
+    if (payload.getType() == Type.STRING) {
+      formatter.appendField("message", payload.getData(), true);
+    } else if (payload.getType() == Type.JSON) {
+      Payload.JsonPayload jsonPayload = (Payload.JsonPayload) payload;
+      formatter.appendJson(jsonPayload.getDataAsMap(), true);
+    } else if (payload.getType() == Type.PROTO) {
```

Comment:
Would it be more right approach to validate `payload.getType() == Type.PROTO` when payload is set? This way it could throw exception on init stage which IMO more correct approach. Also worth mentioning that supported types needs to be documented (e.g. what is supported and what not).

Reply:
Agree about throwing exception earlier although the code reads more clear this way IMO. The method is expected to be used mainly via handlers (JUL or Logback) which do not produce log entries with protoPayload.
I will submit a change tothe [public documentation](https://cloud.google.com/logging/docs/structured-logging) to add a note about this fact.

Reply:
Thank you!